File: src/Particle.cpp
#include "Particle.h"

#include <cmath>

Particle::Particle(const float x, const float y, const float vx, const float vy, const float charge, const float mass)
    : x(x), y(y), vx(vx), vy(vy), charge(charge), mass(mass) {
}

float Particle::distanceTo(const Particle &other) const {
    return static_cast<float>(std::sqrt(std::pow(x - other.x, 2) + std::pow(y - other.y, 2)));
}

sf::Vector2f Particle::calculateForce(const Particle &other) const {
    const float dist = distanceTo(other);
    if (dist == 0) return {0, 0};  // Avoid division by zero

    // Calculate force magnitude
    float forceMagnitude = std::abs(charge * other.charge) / (dist * dist);

    // Calculate the direction of the force
    const float dx = other.x - x;
    const float dy = other.y - y;
    const float distanceNorm = std::sqrt(dx * dx + dy * dy);

    // Normalize direction and apply force magnitude
    float fx = forceMagnitude * (dx / distanceNorm);
    float fy = forceMagnitude * (dy / distanceNorm);

    // Determine if the force is attractive or repulsive
    if (charge * other.charge > 0) {
        // Same charge (repulsive)
        return {-fx, -fy};
    }
    // Opposite charge (attractive)
    return {fx, fy};
}

void Particle::updatePosition(const float dt) {
    x += vx * dt;
    y += vy * dt;
}

void Particle::applyForce(const sf::Vector2f &force, const float dt) {
    vx += (force.x / mass) * dt;
    vy += (force.y / mass) * dt;
}

File: src/main/java/Particle.java
public class Particle {

    private double c; // Charge of the particle (positive or negative)
    private double x, y; // Position on the 2D plane
    private double vx, vy; // Velocity in x and y directions (speed)

    public Particle(double x, double y, double c, double vx, double vy) {
        this.x = x;
        this.y = y;
        this.c = c;
        this.vx = vx;
        this.vy = vy;
    }

    /**
     * Calculates Euclidean distance between two Particles.
     * @param other Other Particle.
     * @return Distance between the two.
     */
    public double distance(Particle other) {
        // TODO: Might wanna optimize this a bit
        return Math.sqrt(Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2));
    }

    /**
     * Calculates the force between two particles.
     * @param other Other particle.
     * @return Force between the two.
     */
    public double force(Particle other) {
        double d = distance(other);
        // Avoiding division by zero
        if (d == 0) return 0;
        return Math.abs(this.c * other.c) / (d * d);
    }

    /**
     * Updates velocity of particle based of the interaction
     * with other particle.
     * @param other Other particle.
     */
    public void interact(Particle other) {
        double d = distance(other);
        if (d == 0) return; // Avoid self-interaction
        double f = force(other);

        double dx = (other.x - this.x) / d;
        double dy = (other.y - this.y) / d;

        // Determine attraction or repulsion
        double direction = (this.c * other.c > 0) ? 1 : -1;

        // Update velocity
        this.vx += direction * f * dx;
        this.vy += direction * f * dy;
    }

    @Override
    public String toString() {
        return "Particle { x=" + x + ", y=" + y + ", vx=" + vx + ", vy=" + vy + ", c=" + c + " }";
    }

}
